### PR TITLE
Validate Conversation ID

### PR DESCRIPTION
**Pull request by aleronupe**
**Proposed changes**:
- Solves issue #7022
- Create a decorator to validate that a conversation ID exists when requests are made to the following endpoints: 
  - GET /conversations/<conversation_id:path>/story
  - POST /conversations/<conversation_id:path>/execute
  - POST /conversations/<conversation_id:path>/predict

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by sara-tagger** [comment]
Thanks for submitting a pull request 🚀 @JustinaPetr will take a look at it as soon as possible ✨

**Comment by wochinge** [comment]
@JustinaPetr I assigned myself as @aleronupe were already taking [here](https://github.com/RasaHQ/rasa/issues/7022#issuecomment-725018825) about it

@aleronupe Will give it a review later today or tomorrow 👍

File: changelog/7022.improvement.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Endpoints that require the existance of the conversation for the specified conversation id now raises an exception.
```
Comment (wochinge) [inline changelog/7022.improvement.md:1]: Can we please list the changed endpoints in here? Ideally also which status code they now return.

File: rasa/server.py [hunk 2]
```diff
@@ -137,6 +137,25 @@ def decorated(*args, **kwargs):
...
 
+def ensure_conversation_exists(app: Sanic):
+    """Wraps a request handler ensuring the conversation exists.
+    """
+
+    def decorator(f):
+        @wraps(f)
+        def decorated(*args, **kwargs):
+            conversation_id = kwargs["conversation_id"]
+            tracker = app.agent.create_processor().get_tracker(conversation_id)
```
Comment (wochinge) [inline rasa/server.py:148]: I'm a bit worried about the performance implications of this as this means that we fetch and de-serialize the tracker twice for these endpoints. What do you think of adding a method `exists` to the `TrackerStore` interface which by default uses `return self.retrieve(conversation_id) is not None` and which we can re-implement with faster versions for each `TrackerStore`?

File: rasa/server.py [hunk 3]
```diff
@@ -137,6 +137,25 @@ def decorated(*args, **kwargs):
     return decorator
 
 
+def ensure_conversation_exists(app: Sanic):
+    """Wraps a request handler ensuring the conversation exists.
+    """
+
+    def decorator(f):
+        @wraps(f)
+        def decorated(*args, **kwargs):
```
Comment (wochinge) [inline rasa/server.py:146]:
```suggestion
        def decorated(*args: Any, **kwargs: Any) -> HTTPResponse:
```

File: rasa/server.py [hunk 4]
```diff
@@ -137,6 +137,25 @@ def decorated(*args, **kwargs):
     return decorator
 
 
+def ensure_conversation_exists(app: Sanic):
```
Comment (wochinge) [inline rasa/server.py:140]:
```suggestion
def ensure_conversation_exists(app: Sanic) -> Callable[..., HTTPResponse]:
```

File: rasa/server.py [hunk 5]
```diff
@@ -137,6 +137,25 @@ def decorated(*args, **kwargs):
...
+    """Wraps a request handler ensuring the conversation exists.
+    """
+
+    def decorator(f):
```
Comment (wochinge) [inline rasa/server.py:144]:
```suggestion
    def decorator(f: Callable[..., HTTPResponse) -> HTTPResponse:
```

File: rasa/server.py (hunk 2 above)
Comment (wochinge) [inline rasa/server.py:148]: There is no need to use the `processor here`. We can directly use
```suggestion
            tracker = app.agent.tracker_store.retrieve(conversation_id)
```

File: rasa/server.py [hunk 6]
```diff
@@ -137,6 +137,25 @@ def decorated(*args, **kwargs):
...
+        @wraps(f)
+        def decorated(*args, **kwargs):
+            conversation_id = kwargs["conversation_id"]
+            tracker = app.agent.create_processor().get_tracker(conversation_id)
+            if tracker is not None:
+                return f(*args, **kwargs)
+            else:
+                raise ErrorResponse(404, "Not found", "Conversation ID not found.")
```
Comment (wochinge) [inline rasa/server.py:152]: let's use a constant for this
```suggestion
                raise ErrorResponse(HTTPStatus.NOT_FOUND, "Not found", "Conversation ID not found.")
```

File: rasa/server.py [hunk 7]
```diff
@@ -776,6 +797,7 @@ async def trigger_intent(request: Request, conversation_id: Text) -> HTTPRespons
     @app.post("/conversations/<conversation_id:path>/predict")
     @requires_auth(app, auth_token)
     @ensure_loaded_agent(app)
+    @ensure_conversation_exists(app)
     async def predict(request: Request, conversation_id: Text) -> HTTPResponse:
         try:
```
Comment (wochinge) [inline rasa/server.py:802]: out of the scope of your PR / optional: We should actually use the `lockstore` here as `processor.predict_next` persists the updated tracker.

**Review by wochinge — COMMENTED** [review]
Thanks for your PR and sorry for the late review! I really appreciate your initiative and contribution 💪 

Can we please also add unit tests for the new changes?

Let me know if I can help somehow!

File: rasa/server.py (hunk 2 above)
Comment (aleronupe) [reply rasa/server.py]: As requested, we've implemented the method `exists` at the `TrackerStore`  interface, so this change ended up as follows:

```python
            if app.agent.tracker_store.exists(conversation_id):
```

**Comment by aleronupe** [comment]
Hello @wochinge, we've just made the required changes and added the unit tests for the new functionalities. Would you please take a look at it?

File: rasa/server.py [hunk 8]
```diff
@@ -149,6 +150,26 @@ def decorated(*args, **kwargs):
...
 
+def ensure_conversation_exists(app: Sanic) -> Callable[..., HTTPResponse]:
+    """Wraps a request handler ensuring the conversation exists.
+    """
+
+    def decorator(f: Callable[..., HTTPResponse]) -> HTTPResponse:
+        @wraps(f)
+        def decorated(*args: Any, **kwargs: Any) -> HTTPResponse:
+            conversation_id = kwargs["conversation_id"]
+            if app.agent.tracker_store.exists(conversation_id):
```
Comment (wochinge) [inline rasa/server.py:161]: You can also extract the agent from the request (which is your first arg) using

```
agent = request.app.agent
```

This allows you to skip passing in `app` needlessly

File: tests/test_server.py [hunk 9]
```diff
@@ -1026,6 +1026,13 @@ async def test_put_tracker(rasa_app: SanicASGITestClient):
     assert events.deserialise_events(evts) == test_events
 
 
+async def test_predict_without_conversation_id(rasa_app: SanicASGITestClient):
+    _, response = await rasa_app.post("/conversations/non_existent_id/predict")
+
+    assert response.status == 404
```
Comment (wochinge) [inline tests/test_server.py:1032]:
```suggestion
    assert response.status == HTTPStatus.NOT_FOUND
```

File: tests/test_server.py [hunk 10]
```diff
@@ -1228,6 +1235,18 @@ async def test_execute(rasa_app: SanicASGITestClient):
...
 
+async def test_execute_without_conversation_id(rasa_app: SanicASGITestClient):
+    # await _create_tracker_for_sender(rasa_app, "test_execute")
+
+    data = {INTENT_NAME_KEY: "utter_greet"}
+    _, response = await rasa_app.post(
+        "/conversations/non_existent_id/execute", json=data
+    )
+
+    assert response.status == 404
```
Comment (wochinge) [inline tests/test_server.py:1246]:
```suggestion
    assert response.status == HTTPStatus.NOT_FOUND
```

File: tests/test_server.py [hunk 11]
```diff
@@ -1513,6 +1532,21 @@ async def test_get_story(
...
+):
+    tracker_store = InMemoryTrackerStore(Domain.empty())
+    monkeypatch.setattr(rasa_app.app.agent, "tracker_store", tracker_store)
+
+    conversation_id = "some-conversation-ID"
+    url = f"/conversations/{conversation_id}/story?"
+
+    _, response = await rasa_app.get(url)
+
+    assert response.status == 404
```
Comment (wochinge) [inline tests/test_server.py:1546]:
```suggestion
    assert response.status == HTTPStatus.NOT_FOUND
```

**Review by wochinge — COMMENTED** [review]
Looks great! Added a few more comments, but nothing critical 👍🏻

**Comment by RomuloSouza** [comment]
Hi @wochinge, we've made the required changes. Can you please take a look?

File: rasa/core/tracker_store.py [hunk 12]
```diff
@@ -173,9 +173,23 @@ def create_tracker(
...
     def save(self, tracker):
-        """Save method that will be overridden by specific tracker"""
+        """Save method that will be overridden by specific tracker."""
         raise NotImplementedError()
 
+    def exists(self, sender_id: Text) -> bool:
+        """Checks if tracker exists for the specified ID.
+
+        This method may be overridden by the specific tracker store for
+        faster implementations
```
Comment (wochinge) [inline rasa/core/tracker_store.py:183]:
```suggestion
        faster implementations.
```

File: rasa/core/tracker_store.py [hunk 13]
```diff
@@ -173,9 +173,23 @@ def create_tracker(
         return tracker
 
     def save(self, tracker):
-        """Save method that will be overridden by specific tracker"""
+        """Save method that will be overridden by specific tracker."""
         raise NotImplementedError()
 
+    def exists(self, sender_id: Text) -> bool:
```
Comment (wochinge) [inline rasa/core/tracker_store.py:179]: we are currently trying to consolidate all the appearances of `sender_id` in the code. Let's make sure new code uses `conversation_id` right away.
```suggestion
    def exists(self, conversation_id: Text) -> bool:
```

File: tests/test_server.py [hunk 14]
```diff
@@ -1463,6 +1470,18 @@ async def test_execute(rasa_app: SanicASGITestClient):
     assert parsed_content["messages"]
 
 
+async def test_execute_without_conversation_id(rasa_app: SanicASGITestClient):
+    # await _create_tracker_for_sender(rasa_app, "test_execute")
+
```
Comment (wochinge) [inline tests/test_server.py:1475]:
```suggestion
```

File: tests/test_server.py [hunk 15]
```diff
@@ -1748,6 +1767,21 @@ async def test_get_story(
     assert response.content.decode().strip() == expected
 
 
+async def test_get_story_without_conversation_id(
+    rasa_app: SanicASGITestClient, monkeypatch: MonkeyPatch
+):
+    tracker_store = InMemoryTrackerStore(Domain.empty())
+    monkeypatch.setattr(rasa_app.app.agent, "tracker_store", tracker_store)
```
Comment (wochinge) [inline tests/test_server.py:1774]: why do we need this?

**Comment by wochinge** [comment]
@aleronupe Can we wrap this up? Happy to help 🙌🏻

File: tests/test_server.py [hunk 16]
```diff
@@ -1748,6 +1765,18 @@ async def test_get_story(
     assert response.content.decode().strip() == expected
 
 
+async def test_get_story_without_conversation_id(
+    rasa_app: SanicASGITestClient, monkeypatch: MonkeyPatch
+):
+    conversation_id = "some-conversation-ID"
+    url = f"/conversations/{conversation_id}/story?"
```
Comment (wochinge) [inline tests/test_server.py:1772]:
```suggestion
    url = f"/conversations/{conversation_id}/story"
```

**Review by wochinge — APPROVED** [review]
💯 thanks a lot @aleronupe

**Comment by wochinge** [comment]
Re the failing windows tests: Windows testing is a bit flakey atm. Re-running the tests should hep 🤞🏻

**Comment by aleronupe** [comment]
@wochinge we've fixed the URL error but windows tests keep failing on a different section of the code. Is there any other change we should make to fix this?

**Comment by wochinge** [comment]
Can you please merge the latest `main` state into your branch? It should be just a matter of re-running / might be fixed on `main` already.

**Comment by aleronupe** [comment]
@wochinge we’ve updated The branch and now all the testes are passing

**Comment by wochinge** [comment]
@aleronupe It's out of date again. Can you somehow make it that I can push to your repo as well? Then i can update the branch as well for you.

**Comment by aleronupe** [comment]
@wochinge Our repo is on an organization from our university, so I think it won’t be possible, but we’ve just updated the branch

**Comment by wochinge** [comment]
Merged 🎉 Thanks for your patience!

**Pull request merged**